### ISSUE TITLE
support for jumpout&jumpnext pairs

### ISSUE THREAD

**Issue by m0ukub1** [windwp/nvim-autopairs#167]
**Is your feature request related to a problem? Please describe.**
Used another autopairs plugin writen in vimscript for a long time, is it possiable support the same feature

**Describe the solution you'd like**
{ | } ... [ ]
--type C-n to jump out
{ } | ... [ ]
--type C-n to jump next
{ } ...[ | ]

**Comment by zmrow** [comment]
[This is another good visualization ](https://github.com/Raimondi/delimitMate/blob/537a1da0fa5eeb88640425c37e545af933c56e1b/doc/delimitMate.txt#L385-L394) of what is being asked for.

At least for me personally, I'd rather not install treesitter just to get this functionality.

In #125 , you mention that it might be possible since you export all the rules.  In #190, @noib3 makes a good point that the plugin could expose `<Plug>autopairs-skip-closing` and users could make their own mapping.  That could solve [any anxiety](https://github.com/windwp/nvim-autopairs/issues/190#issuecomment-991377148) about mapping keys in the plugin itself.

I guess I'm just trying to understand if it is not possible in this plugin, or if it's something you will not add.  No worries either way!  Nice work on this plugin!

**Comment by windwp** [comment]
the problem come from find a match pair if it place on different line and you need treesitter to do that.

if you want to just move a cursor to left you can create a normal mapping to do that or a rule if you want a to check condition.
i don't have a personal use case on that so i will not create that mapping plugautopairs-skip-closing

**Comment by tlindsay** [comment]
I used to use https://github.com/jiangmiao/auto-pairs which provides jumpout behavior like this:
```
Cursor starts here
 │
 │ if (someCondition) {
 └──►|
   }|◄──┐
        │
        │
        Type "}" to jump here
```
Is it possible to configure this plugin to achieve something similar? It's really intuitive to just close whatever pair you're in to jump out.

**Comment by Diomendius** [comment]
@tlindsay I managed to get `nvim-autopairs` to do this via a hack I mentioned [in #147](https://github.com/windwp/nvim-autopairs/issues/147#issuecomment-952639522).

This is a slightly improved version of the code I posted there:

```lua
local nap = require('nvim-autopairs')
local rule = require('nvim-autopairs.rule')
local cond = require('nvim-autopairs.conds')
local utils = require('nvim-autopairs.utils')

local function multiline_close_jump(open, close)
    return rule(close, '')
        :with_pair(function()
            local row, col = utils.get_cursor()
            local line = utils.text_get_current_line()

            if #line ~= col then --not at EOL
                return false
            end

            local unclosed_count = 0
            for c in line:gmatch("[\\" .. open .. "\\" .. close .. "]") do
                if c == open then unclosed_count = unclosed_count + 1 end
                if unclosed_count > 0 and c == close then unclosed_count = unclosed_count - 1 end
            end
            if unclosed_count > 0 then return false end

            local nextrow = row + 1
            if nextrow < vim.api.nvim_buf_line_count(0)
                and vim.regex("^\\s*" .. close):match_line(0, nextrow) then
                return true
            end
            return false
        end)
        :with_move(cond.none())
        :with_cr(cond.none())
        :with_del(cond.none())
        :set_end_pair_length(0)
        :replace_endpair(function()
            return '<esc>xEa'
        end)
end

nap.add_rules {
    multiline_close_jump('(', ')'),
    multiline_close_jump('[', ']'),
    multiline_close_jump('{', '}'),
}
```

IIRC I fixed it preferring to jump over closing braces instead of inserting a closing brace when an unclosed opener exists within another set of braces.

**Comment by tlindsay** [comment]
Amazing! Thanks @Diomendius!

**Comment by vitek-borovsky** [comment]
Trying your solution does it still work for you guys @tlindsay @Diomendius are there any other plugins I need as prerequisits I seem to keep getting error: E5113: Error while calling lua chunk: .../pack/packer/start/nvim-autopairs/lua/nvim-autopairs.lua:115: attempt to index field 'config' (a nil value).
After reading that is a indexing problem i tried nap.add_rule(multiline_close_jump('(', ')')), but it didin't help. Still getting
![image](https://user-images.githubusercontent.com/106488133/190485480-e8ff7e43-3f92-4a0f-bf29-e42733b5449b.png)

TLDR; does it still work?

**Comment by ravener** [comment]
I'd also like to suggest delete functionality, this is done in https://github.com/jiangmiao/auto-pairs

```js
if (true) {| <-- Press backspace here.
}
```
Expected output: 
```js
if (true) 
```

**Comment by amarakon** [comment]
@vitek-borovsky it is not working for me either. This could be an issue with Neovim 0.8.

**Comment by HawkinsT** [comment]
Not nearly as good a solution, but a simple mapping I've been using for a while that will essentially accomplish the same thing and works in 0.8 is:
```lua
vim.keymap.set("i", "<C-l>", "<esc>:exe 'norm! l%%'<CR>a", { silent = true })
```
It's always worked well within paragraphs for me with the two caveats being that you can't really jump between paragraphs easily and it's only one way (you can't jump backwards).

**Comment by brokenbyte** [comment]
My take on this:
```lua
local function multiline_close_jump(open, close)
    return rule(close, "")
        :with_pair(function()
            local row, col = utils.get_cursor(0)
            local line = utils.text_get_current_line(0)

            if #line ~= col then --not at EOL
                return false
            end

            local unclosed_count = 0
            for c in line:gmatch("[\\" .. open .. "\\" .. close .. "]") do
                if c == open then unclosed_count = unclosed_count + 1 end
                if unclosed_count > 0 and c == close then unclosed_count = unclosed_count - 1 end
            end
            if unclosed_count > 0 then return false end

            local nextrow = row + 1

            if nextrow < vim.api.nvim_buf_line_count(0) and vim.regex("^\\s*" .. close):match_line(0, nextrow) then
                return true
            end
            return false
        end)
        :with_move(cond.none())
        :with_cr(cond.none())
        :with_del(cond.none())
        :set_end_pair_length(0)
        :replace_endpair(function(opts)
            local row, _col = utils.get_cursor(0)
            local action = vim.regex("^" .. close):match_line(0, row + 1) and "a" or ("0f%sa"):format(opts.char)
            return ("<esc>xj%s"):format(action)
        end)
end
```

@camilledejoye 's version _almost_ works for me, but doesn't leave me in insert mode when the matching `}` is in the first column for some reason. E.g.:

```lua
local t = {
  y = {
    |
  }
}
```

works fine and results in

```lua
local t = {
  y = {
  }|
}
```

but pressing `}` again right away puts my cursor on the final `}` and leaves me in normal mode:

```lua
local t = {
  y = {
  }
}| -- Cursor here, but in Normal mode, not Insert
```

After playing around with different variations for `replace_endpair` return values, it seems to have something to do with using `f{` while in the first column or something around there, because it's only an issue when the matching pair is the first character on the line afaict; even adding a space before it works properly.

My fix just checks if the first character on the line is the matching pair and runs `<esc>a`, otherwise does `<esc>0f}a`.
Also note that I don't have it delete blank lines inside the matching pairs as I use the blank line as a reminder to put something there later, so you'll need to change the string if you want to keep that.

**Comment by rish987** [comment]
To get back to the original question, I have put together this quick & dirty solution for using a single key to jump over multiple nested delimiters, though it only works in the single line case:

```lua
local check_next_delim = function(line, col)
  for _, rule in ipairs(npairs.get_buf_rules(vim.api.nvim_get_current_buf())) do
    if rule.start_pair then
      local delim = rule.end_pair
      if line:find(delim, col, true) == col then return delim end
    end
  end
end

local skip_delims = function()
  local col = vim.fn.col('.')
  local line = vim.fn.getline('.')

  local next_delim = check_next_delim(line, col)
  while next_delim do
    for _ = 1, vim.fn.strdisplaywidth(next_delim) do
      vim.fn.feedkeys(vim.api.nvim_replace_termcodes("<Right>", true, true, true))
    end

    col = col + #next_delim
    if col > #line then return end

    next_delim = check_next_delim(line, col)
  end
end

vim.keymap.set("i", "<Tab>", skip_delims)
```

**Comment by rodhash** [comment]
+1 miss a LOT the fly mode.. that's so good

Any of these "fly mode" hack still work?

It seems this isn't working on latest nvim versions, I'm on `v0.11.0-dev-301` and the fly mode isn't working. For me it's working only the standard nvim-autopairs behavior.

I guess meanwhile I'll stick with the old school vimL plugin.. tried these hack but for me it still lack the fly mode
